### ISSUE TITLE
Question: How to profile an Orleans application?

### ISSUE THREAD

**Issue by philippecp** [dotnet/orleans#1593]
I'm trying to profile a silo but I'm having a very hard time getting relevant data. Visual Studio profiler reports most time taken in BlockingCollection<T>.Take (which I'm assuming is Orleans Task Scheduler). Other than that I get a lot of time spent in unknown code and the async/task based nature of Orleans seems to confuse the profiler and I can't get stack trace that shows me which one of my grain functions is taking up all the time. The only thing I've come up with by trial and error is that grain communication seems very expensive even between stateless workers so I've had to turn many grain classes into regular classes and make sure to make chunky calls on remaining ones.

So in short, how does one go about finding bottlenecks?

**Comment by sergeybykov** [comment]
Profiling isn't straightforward, indeed. Before running a profiler we recommend to make sure ServerGC is enable and `Immutable<>` is leveraged where possible. Then look at messaging statistics. As you mentioned, even a local grain call is more expensive than a direct object call. A remote call even more so. Messaging statistics are included in the logs, by default every 5 minutes. Elimination of chatty communications and optimization of data structures being sent between grains are usually first order optimizations.

Also, make sure you are profiling under heavy load, ideally at full utilization, so that spin locks don't skew the numbers too much.

**Comment by philippecp** [comment]
I think the messaging statistics would be very helpful! Is there a way way to trigger flushing statistics before timer expires?

I also have some general questions:

1. Is a stateless worker grain also re-entrant?
2. Is there a significant performance difference between using Immutable<T> vs ImmutableAttribute? I find using the attribute easier because when I make concurrent calls with Immutable<T> return values I have to transform result of Task.WhenAll() to get Immutable<T>.Value for array.
3. When you say optimizing data structures being sent between grains, do you mean optimize for message size or actual implementation (ex: It's faster to send arrays than lists). Does using ms Bond serializer offer better performance than default serializer?

Thanks!

**Comment by sergeybykov** [comment]
>Is there a way way to trigger flushing statistics before timer expires?

I'm not sure I understand the question. Do you mean if it's possible to report statistics more often than every 5 minutes? If that's the questions, then yes, the reporting period is configurable via `StatisticsLogWriteInterval` - https://github.com/dotnet/orleans/blob/master/src/Orleans/Configuration/NodeConfiguration.cs#L190.

>Is a stateless worker grain also re-entrant.

Not by default, but can be.

>Is there a significant performance difference between using Immutable vs ImmutableAttribute?

There shouldn't be any noticeable difference.

>When you say optimizing data structures being sent between grains, do you mean optimize for message size or actual implementation (ex: It's faster to send arrays than lists).

In general, send smaller simpler data structures, and avoid sending unnecessary data. I'm not sure about arrays vs. lists. Probably arrays are more efficient due to fewer allocations, but it depends on the size. I wouldn't worry about such micro-optimizations upfront though. The bigger fish to fry is usually chatty protocols and unnecessary data that can be easily cached.

>Does using ms Bond serializer offer better performance than default serializer?

Hard to tell without measuring a specific case. The Bond serializer makes the most sense if your data already comes or is stored in the Bond format. Orleans serializer is schema-less, so it's easier to use, but it doesn't support versioning yet. I would start with the Orleans serializer unless your data is already in Bond.
